src/StackExchange.Redis/Configuration/LoggingTunnel.cs

```diff
@@ -0,0 +1,624 @@
+
+        long count = 0;
+        while (true)
+        {
+            var sent = await ReadOneAsync(outPipe, arena, isInbound: false);
+            ContextualRedisResult received;
+            try
+            {
+                do
+                {
+                    received = await ReadOneAsync(inPipe, arena, isInbound: true);
+                    if (received.IsOutOfBand && received.Result is not null)
+                    {
+                        // spoof an empty request for OOB messages
+                        pair(RedisResult.NullSingle, received.Result);
+                    }
+                } while (received.IsOutOfBand);
+            }
+            catch (Exception ex)
+            {
+                // if we got an exception following a command, spoof that as a pair,
+                // so we see the message that had a corrupted reply
+                if (sent.Result is not null)
+                {
+                    pair(sent.Result, RedisResult.Create(ex.Message, ResultType.Error));
+                }
+                throw; // still surface the original exception
+            }
+
+            if (sent.Result is null || received.Result is null) break; // no more paired messages
+
+            pair(sent.Result, received.Result);
+            count++;
+        }
+        return count;
+    }
+
+
+    /// <summary>
+    /// Replay the RESP messages all the streams in a folder, invoking a callback per operation
+    /// </summary>
+    public static async Task<long> ReplayAsync(string path, Action<RedisResult, RedisResult> pair)
+    {
+        long total = 0;
+        foreach (var outPath in Directory.EnumerateFiles(path, "*.out"))
+        {
+            var inPath = Path.ChangeExtension(outPath, "in");
+            if (!File.Exists(outPath)) continue;
+
+            using var outFile = File.OpenRead(outPath);
+            using var inFile = File.OpenRead(inPath);
+            total += await ReplayAsync(outFile, inFile, pair);
+        }
+        return total;
+    }
+
+    private static async ValueTask<ContextualRedisResult> ReadOneAsync(PipeReader input, Arena<RawResult> arena, bool isInbound)
+    {
+        while (true)
+        {
+            var readResult = await input.ReadAsync().ForAwait();
+            var buffer = readResult.Buffer;
+            int handled = 0;
+            var result = buffer.IsEmpty ? default : ProcessBuffer(arena, ref buffer, isInbound);
+            input.AdvanceTo(buffer.Start, buffer.End);
+
+            if (result.Result is not null) return result;
+
+            if (handled == 0 && readResult.IsCompleted)
+            {
+                break; // no more data, or trailing incomplete messages
+            }
+        }
+        return default;
+    }
+
+    /// <summary>
+    /// Validate a RESP stream and return the number of top-level RESP fragments.
+    /// </summary>
+    public static async Task<long> ValidateAsync(string path)
```

Comment:
Add <param> doc comment about how you can pass a path to a file OR directory?

Reply:
fair; 5ac13be3a5ad1f06ed725ebd5bbd8fcf18324476